### PR TITLE
Refactor ConsumerEvent and related modules

### PR DESCRIPTION
**Pull request by acco**
This pull request refactors the consumer event and record data structures to improve code organization and maintainability. The changes include:

1. Extracted `ConsumerEventData` and `ConsumerMessageMetadata` into separate modules.
2. Created a new `ConsumerRecordData` module to represent record data.
3. Updated the `ConsumerEvent` and `ConsumerRecord` modules to use the new data structures.
4. Modified the `Sequin.Map.from_ecto/2` function to accept an optional `keep_nils` parameter.
5. Updated factory functions in `ConsumersFactory` to work with the new data structures.
6. Added new factory functions for `ConsumerMessageMetadata`, `ConsumerEventData`, and `ConsumerRecordData`.

These changes enhance the modularity of the codebase and provide better separation of concerns for consumer-related data structures.

### CODE REVIEW

**Comment by acco** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/118?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#118** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/118?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#117** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/117?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/116?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#115** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/115?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#114** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/114?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#113** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/113?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#112** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/112?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#111** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/111?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#110** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/110?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @acco and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/sequin/consumers/consumer_event.ex [hunk 1]
```diff
@@ -5,42 +5,10 @@ defmodule Sequin.Consumers.ConsumerEvent do
   import Ecto.Changeset
   import Ecto.Query
 
+  alias Sequin.Consumers.ConsumerEventData
+  alias Sequin.Consumers.ConsumerMessageMetadata
   alias Sequin.Streams.ConsumerEventWithConsumerInfo
 
-  defmodule Data do
```
Comment (acco) [inline lib/sequin/consumers/consumer_event.ex:10]: Figured we might as well move out, as the plan was to share metadata between the two.

However I'm now wondering if the metadata will be different as well. I didn't mind it in the same file, so we can always move it back in

I do like calling it `Consumers.ConsumerEventData` though - I like flat

File: lib/sequin/map.ex [hunk 2]
```diff
@@ -41,18 +41,22 @@ defmodule Sequin.Map do
...
     struct
     |> Map.from_struct()
-    |> from_ecto()
+    |> from_ecto(opts)
   end
 
   # May be a struct that was recently converted into a map
-  def from_ecto(map) do
+  def from_ecto(map, opts) do
+    keep_nils = Keyword.get(opts, :keep_nils, false)
```
Comment (acco) [inline lib/sequin/map.ex:55]: `from_ecto` drops nils. We don't always want that behavior.

Example: When doing a controller test, we don't want the client to send in a bunch of nil keys

But when testing ConsumerEventData, if `changes` is `nil`, we don't want to drop it

In our factories, we could have two kinds of `_attrs` functions for each object, but as you'll see it didn't really come to that -- there's just certain factories that we don't want to drop nils for

File: test/support/factory/consumers_factory.ex [hunk 3]
```diff
@@ -217,14 +220,63 @@ defmodule Sequin.Factory.ConsumersFactory do
   defp generate_value(:float), do: Factory.float()
   defp generate_value(:boolean), do: Factory.boolean()
 
+  # HttpEndpoint
```
Comment (acco) [inline test/support/factory/consumers_factory.ex:223]: moved

File: lib/sequin/consumers/consumer_message_metadata.ex [hunk 4]
```diff
@@ -0,0 +1,21 @@
+defmodule Sequin.Consumers.ConsumerMessageMetadata do
```
Comment (RTLS) [inline lib/sequin/consumers/consumer_message_metadata.ex:1]: so `Message` here refers generically to an event or a record 🤔 are we going to use that for any functions that generically can get events or records?

Comment (RTLS) [reply lib/sequin/consumers/consumer_message_metadata.ex:1]: `ConsumerMetadata` doesn't really work though 🤔

**Review by RTLS — APPROVED** [review]

File: lib/sequin/consumers/consumer_message_metadata.ex (hunk 4 above)
Comment (acco) [reply lib/sequin/consumers/consumer_message_metadata.ex:1]: That's right - and that is my struggle! We us messages a couple more times in future

I think the ConsumerMetadata may diverge more in future

**Pull request merged**